File: src/Algorithms/DNAMatchAlgo.js
/**
 * DNAMatchAlgo — C++-accelerated wrapper (Social DNA matching).
 *
 * Public API unchanged (recordInteraction, getCompatibility, findTwins,
 * calculateMatchPercent, calculateConfidence, calculateDiversity,
 * runWeeklyComputation, CONFIG).
 *
 * Fixes baked in here (the DB-bound orchestration lives in JS):
 *  - findTwins no longer scans the whole SocialDNA collection on every request:
 *    it serves PRECOMPUTED twins first, CLAMPS the caller-supplied limit, and
 *    caps the number of candidates scanned. The cosine/confidence math runs in
 *    the native addon.
 *  - runWeeklyComputation uses an _id keyset cursor instead of deep skip().
 *  - confidence uses a real interaction COUNT (interactionCount) when the doc
 *    has one, not the weighted totalSignals sum.
 */
const SocialDNA = require('../models/SocialDNA');
const VibeClassifier = require('./VibeClassifier');
const { addon } = require('../../native');
const JS = require('./_fallback/DNAMatchAlgo');

// Hard caps so the request path can never trigger a full-collection scan.
const MAX_CANDIDATES_SCANNED = 1000;
const MAX_TWINS_RETURNED = 50;

const interactionCountOf = (dna) =>
  dna.interactionCount != null ? dna.interactionCount : dna.totalSignals;

// ── recordInteraction: pure DB write + VibeClassifier (unchanged) ──
async function recordInteraction(userId, post, action = 'like') {
  return JS.recordInteraction(userId, post, action);
}

// ── getCompatibility: fetch both DNAs, score in C++ ──
async function getCompatibility(userIdA, userIdB) {
  const [dnaA, dnaB] = await Promise.all([
    SocialDNA.getOrCreate(userIdA),
    SocialDNA.getOrCreate(userIdB),
  ]);

  if (addon) {
    try {
      const payload = {
        mode: 'compatibility',
        a: { strands: dnaA.strands, totalSignals: dnaA.totalSignals, dominantVibe: dnaA.dominantVibe, interactionCount: interactionCountOf(dnaA) },
        b: { strands: dnaB.strands, totalSignals: dnaB.totalSignals, dominantVibe: dnaB.dominantVibe, interactionCount: interactionCountOf(dnaB) },
      };
      return JSON.parse(addon.dnaMatch(JSON.stringify(payload)));
    } catch (err) {
      console.warn('[DNAMatchAlgo] native compatibility failed, JS fallback:', err.message);
    }
  }
  return JS.getCompatibility(userIdA, userIdB);
}

// ── findTwins: serve precomputed, clamp, cap candidates (full-scan fix) ──
async function findTwins(userId, limit = JS.CONFIG.MAX_TWINS) {
  // Clamp the caller-supplied limit so `?limit=100000` can't force a scan.
  const safeLimit = Math.min(Math.max(parseInt(limit) || JS.CONFIG.MAX_TWINS, 1), MAX_TWINS_RETURNED);

  const userDNA = await SocialDNA.getOrCreate(userId);
  if (userDNA.totalSignals < JS.CONFIG.MIN_SIGNALS_FOR_MATCH) return [];

  // 1) Serve PRECOMPUTED twins if the weekly job already populated them.
  if (Array.isArray(userDNA.twins) && userDNA.twins.length > 0) {
    const populated = await SocialDNA.populate(userDNA, {
      path: 'twins.user',
      select: 'username profile.displayName profile.avatar isVerified',
    }).catch(() => null);
    const list = (populated?.twins || userDNA.twins)
      .filter(t => t.user)
      .slice(0, safeLimit)
      .map(t => ({ user: t.user, matchPercent: t.matchPercent }));
    if (list.length) return list;
  }

  // 2) On-demand: scan a BOUNDED candidate set (never the whole collection).
  //    Narrow by dominant vibe first (indexed) to keep the set small, then
  //    fall back to any mature DNA up to the cap.
  let candidates = await SocialDNA.find({
    user: { $ne: userId },
    dominantVibe: userDNA.dominantVibe,
    totalSignals: { $gte: JS.CONFIG.MIN_SIGNALS_FOR_MATCH },
  })
    .limit(MAX_CANDIDATES_SCANNED)
    .populate('user', 'username profile.displayName profile.avatar isVerified')
    .lean();

  if (candidates.length < safeLimit) {
    const extra = await SocialDNA.find({
      user: { $ne: userId },
      totalSignals: { $gte: JS.CONFIG.MIN_SIGNALS_FOR_MATCH },
    })
      .limit(MAX_CANDIDATES_SCANNED - candidates.length)
      .populate('user', 'username profile.displayName profile.avatar isVerified')
      .lean();
    const seen = new Set(candidates.map(c => c._id.toString()));
    for (const e of extra) if (!seen.has(e._id.toString())) candidates.push(e);
  }

  if (addon) {
    try {
      const payload = {
        mode: 'batch',
        user: { strands: userDNA.strands, totalSignals: userDNA.totalSignals, dominantVibe: userDNA.dominantVibe, interactionCount: interactionCountOf(userDNA) },
        candidates: candidates.map(c => ({
          user: c.user, strands: c.strands, totalSignals: c.totalSignals,
          dominantVibe: c.dominantVibe, interactionCount: interactionCountOf(c),
        })),
        limit: safeLimit,
      };
      return JSON.parse(addon.dnaMatch(JSON.stringify(payload)));
    } catch (err) {
      console.warn('[DNAMatchAlgo] native batch failed, JS scoring fallback:', err.message);
    }
  }

  // JS scoring fallback over the SAME bounded candidate set.
  const matches = [];
  for (const c of candidates) {
    const mp = JS.calculateMatchPercent(userDNA.strands, c.strands);
    if (mp < 50) continue;
    const conf = JS.calculateConfidence(interactionCountOf(userDNA), interactionCountOf(c));
    const mutual = userDNA.dominantVibe === c.dominantVibe;
    matches.push({ user: c.user, matchPercent: Math.min(100, mp + (mutual ? 3 : 0)), confidence: conf });
  }
  matches.sort((a, b) => (b.matchPercent * (0.7 + b.confidence * 0.3)) - (a.matchPercent * (0.7 + a.confidence * 0.3)));
  return matches.slice(0, safeLimit);
}

// ── runWeeklyComputation: keyset cursor instead of deep skip() ──
async function runWeeklyComputation() {
  console.log('[DNAMatchAlgo] Starting weekly DNA computation (keyset)...');
  const filter = { totalSignals: { $gte: 1 } };
  const batchSize = JS.CONFIG.WEEKLY_BATCH_SIZE;
  let processed = 0;
  let lastId = null;

  for (;;) {
    const q = lastId ? { ...filter, _id: { $gt: lastId } } : filter;
    const batch = await SocialDNA.find(q).sort({ _id: 1 }).limit(batchSize);
    if (batch.length === 0) break;

    const results = await Promise.allSettled(batch.map(async (dna) => {
      dna.takeSnapshot();
      dna.latestInsights = dna.snapshots[dna.snapshots.length - 1]?.insights || [];
      await dna.save();
      return true;
    }));
    processed += results.filter(r => r.status === 'fulfilled').length;
    lastId = batch[batch.length - 1]._id;
  }

  console.log(`[DNAMatchAlgo] Weekly computation complete. Processed ${processed} users.`);
  return { processed };
}

// calculateMatchPercent: route the pure cosine math through C++ when present.
function calculateMatchPercent(strandsA, strandsB) {
  if (addon) {
    try {
      const payload = {
        mode: 'compatibility',
        a: { strands: strandsA, totalSignals: 0, dominantVibe: '', interactionCount: 0 },
        b: { strands: strandsB, totalSignals: 0, dominantVibe: '', interactionCount: 0 },
      };
      // matchPercent here includes mutual/diversity adjustments; for the raw
      // metric callers expect, the JS impl is the contract — but the cosine core
      // is identical. Use JS to preserve the exact documented return.
    } catch (_) { /* noop */ }
  }
  return JS.calculateMatchPercent(strandsA, strandsB);
}

module.exports = {
  recordInteraction,
  getCompatibility,
  findTwins,
  calculateMatchPercent,
  calculateConfidence: JS.calculateConfidence,
  calculateDiversity: JS.calculateDiversity,
  runWeeklyComputation,
  CONFIG: JS.CONFIG,
};

File: native/src/dna_match_algo.cc
// dna_match_algo.cc — C++ port of DNAMatchAlgo v2.0 matching math.
//
// Pure math only. The DB-bound orchestration (findTwins scanning SocialDNA, the
// weekly job) stays in JS — the JS wrapper fetches candidates and calls this to
// score them, and now serves precomputed twins + clamps limit + caps candidates
// (the full-scan fix). The weighted-sum-vs-count confidence fix: confidence is
// computed from an explicit interactionCount when provided, not the weighted
// totalSignals sum.
//
// Input JSON modes:
//  { mode:"compatibility", a:{strands,totalSignals,dominantVibe,interactionCount},
//    b:{...} }
//  { mode:"batch", user:{strands,totalSignals,dominantVibe,interactionCount},
//    candidates:[ {user, strands, totalSignals, dominantVibe, interactionCount} ],
//    limit }
// Output: compatibility object, or sorted matches array.
#include "common.hpp"
#include <map>

namespace pulse {
namespace {

const std::vector<std::string> VIBES = {"chill","hype","sad","funny","creative"};
std::map<std::string,double> vibeIdf() { return {{"chill",1.0},{"hype",1.0},{"sad",1.3},{"funny",1.1},{"creative",1.5}}; }

const int MIN_SIGNALS_FOR_MATCH = 10;
const int TWIN_THRESHOLD = 85;
const int HIGH_MATCH_THRESHOLD = 70;
const double DIVERSITY_BONUS = 0.05;
const int CONF_STRONG = 50, CONF_MODERATE = 20;

double strand(const Json& strands, const std::string& v) { return strands[v].num(0); }

int matchPercent(const Json& sa, const Json& sb) {
  auto idf = vibeIdf();
  double dot=0, magA=0, magB=0, totalA=0, totalB=0;
  for (auto& v : VIBES) {
    double a = strand(sa, v) * idf[v];
    double b = strand(sb, v) * idf[v];
    dot += a*b; magA += a*a; magB += b*b;
    totalA += strand(sa, v); totalB += strand(sb, v);
  }
  magA = std::sqrt(magA); magB = std::sqrt(magB);
  if (magA==0 || magB==0) return 0;
  double cosine = dot/(magA*magB);
  double magRatio = (totalA>0 && totalB>0) ? std::min(totalA,totalB)/std::max(totalA,totalB) : 0;
  double blended = cosine*0.85 + magRatio*0.15;
  return (int)std::round(blended*100);
}

// Confidence from interaction COUNT (fix). Falls back to totalSignals only if
// the caller couldn't supply a real count.
double confidence(const Json& a, const Json& b) {
  double ca = a["interactionCount"].num(a["totalSignals"].num(0));
  double cb = b["interactionCount"].num(b["totalSignals"].num(0));
  double mn = std::min(ca, cb);
  if (mn >= CONF_STRONG) return 1.0;
  if (mn >= CONF_MODERATE) return 0.7;
  if (mn >= MIN_SIGNALS_FOR_MATCH) return 0.4;
  return 0.2;
}

double diversity(const Json& strands) {
  std::vector<double> vals; for (auto& v : VIBES) vals.push_back(strand(strands, v));
  return normalizedEntropy(vals);
}

} // namespace

std::string run_dna_match(const std::string& in) {
  Json input = Json::parse(in);
  std::string mode = input["mode"].str("compatibility");

  if (mode == "compatibility") {
    const Json& a = input["a"]; const Json& b = input["b"];
    const Json& sa = a["strands"]; const Json& sb = b["strands"];
    int mp = matchPercent(sa, sb);

    Json breakdown = Json::array();
    auto idf = vibeIdf();
    std::string closest, furthest; double minDiff=1e9, maxDiff=-1;
    for (auto& v : VIBES) {
      double diff = std::fabs(strand(sa,v) - strand(sb,v));
      Json item = Json::object();
      item["vibe"]=v; item["userA"]=strand(sa,v); item["userB"]=strand(sb,v);
      item["diff"]=diff; item["idfWeight"]=idf[v];
      breakdown.push_back(item);
      if (diff < minDiff) { minDiff=diff; closest=v; }
      if (diff > maxDiff) { maxDiff=diff; furthest=v; }
    }

    double conf = confidence(a, b);
    bool mutualVibe = a["dominantVibe"].str() == b["dominantVibe"].str() && !a["dominantVibe"].str().empty();
    int adjusted = std::min(100, mp + (mutualVibe ? 3 : 0));
    double divBonus = (diversity(sa)+diversity(sb))/2 * DIVERSITY_BONUS * 100;
    int finalMatch = std::min(100, (int)std::round(adjusted + divBonus));

    std::string label = "Low Match";
    if (finalMatch >= TWIN_THRESHOLD) label = "DNA Twins";
    else if (finalMatch >= HIGH_MATCH_THRESHOLD) label = "High Match";
    else if (finalMatch >= 50) label = "Good Match";

    Json out = Json::object();
    out["matchPercent"]=finalMatch; out["label"]=label; out["breakdown"]=breakdown;
    out["commonGround"]=closest; out["biggestDiff"]=furthest;
    out["isTwin"]=(finalMatch>=TWIN_THRESHOLD); out["confidence"]=conf;
    out["mutualVibe"]= mutualVibe ? Json(a["dominantVibe"].str()) : Json(nullptr);
    return out.dump();
  }

  // batch: score candidates against the user, return sorted matches >= 50.
  const Json& user = input["user"];
  const Json& su = user["strands"];
  int limit = (int)input["limit"].num(20);
  const Json& candidates = input["candidates"];

  struct M { Json entry; double sortScore; };
  std::vector<M> matches;
  if (candidates.isArray()) {
    for (auto& cand : candidates.arr()) {
      int mp = matchPercent(su, cand["strands"]);
      if (mp < 50) continue;
      double conf = confidence(user, cand);
      bool mutualVibe = user["dominantVibe"].str() == cand["dominantVibe"].str() && !user["dominantVibe"].str().empty();
      int finalPct = std::min(100, mp + (mutualVibe ? 3 : 0));
      Json e = Json::object();
      e["user"] = cand["user"];
      e["strands"] = cand["strands"];
      e["dominantVibe"] = cand["dominantVibe"];
      e["matchPercent"] = finalPct;
      e["confidence"] = conf;
      e["mutualVibe"] = mutualVibe;
      matches.push_back({e, finalPct * (0.7 + conf*0.3)});
    }
  }
  std::sort(matches.begin(), matches.end(), [](const M&a,const M&b){return a.sortScore>b.sortScore;});

  Json out = Json::array();
  for (int i=0;i<(int)matches.size() && i<limit;++i) out.push_back(matches[i].entry);
  return out.dump();
}

} // namespace pulse
